### PR TITLE
upgrade collective fleet api

### PR DESCRIPTION
**Pull request by guru4elephant**
test=develop
```python
from paddle.fluid.incubate.fleet.collective import fleet, DistributedStrategy
import paddle.fluid.incubate.fleet.base.role_maker as role_maker

# total loss definination

role = role_maker.PaddleCloudRoleMaker(is_collective=True)
fleet.init(role)
strategy = DistributedStrategy()
optimizer = fluid.optimizer.SGD(learning_rate=0.001)
optimizer = fleet.distributed_optimizer(optimizer, strategy=strategy)
opts, param_and_grads = optimizer.minimize(total_loss)

```

### CODE REVIEW

File: python/paddle/fluid/incubate/fleet/collective/__init__.py [hunk 1]
```diff
@@ -231,32 +165,51 @@ def minimize(self,
...
+        t.transpile(
+            trainer_id=fleet.worker_index(),
+            trainers=fleet.worker_endpoints(),
+            current_endpoint=fleet.worker_endpoints()[fleet.worker_index()],
+            startup_program=startup_program,
+            program=loss.block.program)
+        return opts, param_and_grads
+
+
+class FP32SGDOptimizer(CollectiveOpBasedOptimizer):
```
Comment (gavin1332) [inline python/paddle/fluid/incubate/fleet/collective/__init__.py:191]: GradAllreduce is not only suitable for SGD but most gradient-based optimization, isn't it?

**Review by gavin1332 — COMMENTED** [review]
GradAllreduce is not only suitable for SGD but most other gradient-based optimizations, isn't it?

File: python/paddle/fluid/incubate/fleet/collective/__init__.py (hunk 1 above)
Comment (guru4elephant) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:191]: of course. It is only a symbol for inner usage and is not visible to users. SGDOptimizer is not meant to be the sgd optimizer. Similarly, I don't think localsgd only works with sgd optimizer.

Comment (gavin1332) [reply python/paddle/fluid/incubate/fleet/collective/__init__.py:191]: That's OK

**Review by gavin1332 — APPROVED** [review]
LGTM
